drud/ddev
# Inform users when a `ddev` update is available

## The Problem:
There is a need for users to be notified of updates to ddev. 

## The Fix:
This code introduces periodic (weekly) checking of github releases. Using semver comparisons it informs the user if there are available updates by linking them to the release URL.

Overview:
* Update checks are weekly
* Currently there is no major/minor/patch release versioning or anything along those lines. We may want to add that in the future, but I wanted to keep things simple for now. If there's a newer release, you'll get a weekly message informing you of the release.
* Updates do not run if it's not an official release. i.e. dev builds with -dirty or commitish as the current version do not get updates ran. This is currently a very naive check which just looks for a "-" in the version and doesn't run the update if it exists.
* A file is stored at ~/.ddev/.update. File modification date is used to track time since last update.
* The package which controls this is fairly general. I believe it could be used for any of our projects, as you just give it a repository and current version and it lets you know if an update is available, and what the URL for the update is. 

## The Test:

The easiest way to test is the following:
- Build this binary with an old version. For instance, `make DdevVersion=v0.2 darwin`
- Run any ddev command, such as `ddev version`
- The first run you should see information about an available update. Subsequent runs should not have it (until a week has passed)
- You can alter the file modification date using the touch command. The following command modifies it (on OSX) to May 1st 2017 at 01:01 AM. `touch -mt 201705010101  ~/.ddev/.update`
- After altering the modification time you should see the update message again. An alteration of the file modification date to be less than a week old should not trigger a new update message.


## Automation Overview:
Please see `updatecheck_test.go` in the `updatecheck` package.

## Related Issue Link(s):
The parent issue for this is at https://github.com/drud/ddev/issues/99
